# Проект по SQL
Описание: Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

# Загрузка библиотек и подключение к базе

In [1]:
import pandas as pd 
from sqlalchemy import create_engine 

In [2]:

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', 
             'port': 6432,# порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],  
                                                         db_config['pwd'],  
                                                         db_config['host'], 
                                                         db_config['port'],  
                                                         db_config['db']) 
 

In [3]:
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Выгрузка таблиц

In [4]:
list_table = ['books','authors','publishers','ratings','reviews']

In [17]:
for table in list_table:
    query = f"""SELECT 
        *
    FROM 
        {table};

    """
    df= pd.io.sql.read_sql(query, con = engine)
    display(f'Название таблицы: {table}')
    display(df.info())
    display(df.head())
    display('--------------------------------------------------------')

'Название таблицы: books'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
book_id             1000 non-null int64
author_id           1000 non-null int64
title               1000 non-null object
num_pages           1000 non-null int64
publication_date    1000 non-null object
publisher_id        1000 non-null int64
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

'--------------------------------------------------------'

'Название таблицы: authors'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
author_id    636 non-null int64
author       636 non-null object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

'--------------------------------------------------------'

'Название таблицы: publishers'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
publisher_id    340 non-null int64
publisher       340 non-null object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

'--------------------------------------------------------'

'Название таблицы: ratings'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
rating_id    6456 non-null int64
book_id      6456 non-null int64
username     6456 non-null object
rating       6456 non-null int64
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

'--------------------------------------------------------'

'Название таблицы: reviews'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
review_id    2793 non-null int64
book_id      2793 non-null int64
username     2793 non-null object
text         2793 non-null object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

'--------------------------------------------------------'

## Посчитайте, сколько книг вышло после 1 января 2000 года;

In [6]:
query = """
SELECT 
    COUNT(*) AS count_books
FROM 
    books
WHERE 
    publication_date >= '2000-01-01';

"""

pd.io.sql.read_sql(query, con = engine)

count_books
0          821

После 1.01.2000 года вышло 821 книг

## Для каждой книги посчитайте количество обзоров и среднюю оценку

In [7]:
query = """
SELECT 
    books.book_id AS book_id,
    books.title AS title_book,
    subq_1.rating_avg,
    subq.review_count
FROM
    books
INNER JOIN
    (
    SELECT 
        ratings.book_id AS book_id,
        AVG(ratings.rating) AS rating_avg
    FROM
        ratings
    GROUP BY
        ratings.book_id ) AS subq_1
ON books.book_id = subq_1.book_id  
INNER JOIN
    (
    SELECT 
        reviews.book_id AS book_id,
        COUNT(reviews.review_id) AS review_count
    FROM
        reviews
    GROUP BY
        reviews.book_id
        ) AS subq      
ON books.book_id = subq.book_id
ORDER BY
    subq_1.rating_avg DESC,
    subq.review_count DESC
LIMIT
    10;
"""
pd.io.sql.read_sql(query, con = engine)

book_id                                         title_book  rating_avg  \
0       17                      A Dirty Job (Grim Reaper  #1)         5.0   
1      553            School's Out—Forever (Maximum Ride  #2)         5.0   
2      444       Moneyball: The Art of Winning an Unfair Game         5.0   
3      224                                      Evening Class         5.0   
4      418                                              March         5.0   
5      390                                    Light in August         5.0   
6      972  Wherever You Go  There You Are: Mindfulness Me...         5.0   
7       57                    Act of Treason (Mitch Rapp  #9)         5.0   
8      182                                         Dead Souls         5.0   
9      699  The Demon-Haunted World: Science as a Candle i...         5.0   

   review_count  
0             4  
1             3  
2             3  
3             2  
4             2  
5             2  
6             2  
7             2  
8             2  
9             2

Лидером по средней оценке и количеству обзоров является A Dirty Job (Grim Reaper #1)

## Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
 

In [8]:
query = """
SELECT 
    publishers.publisher AS publisher,
    COUNT(books.num_pages) AS books_cnt
FROM 
    books

INNER JOIN publishers ON publishers.publisher_id = books.publisher_id
WHERE
    num_pages >= 50
GROUP BY 
    publishers.publisher
ORDER BY 
    books_cnt DESC
LIMIT
    10;
"""
pd.io.sql.read_sql(query, con = engine)

publisher  books_cnt
0              Penguin Books         42
1                    Vintage         31
2   Grand Central Publishing         25
3           Penguin Classics         24
4           Ballantine Books         19
5                     Bantam         19
6                    Berkley         17
7         St. Martin's Press         14
8              Berkley Books         14
9  William Morrow Paperbacks         13

Издательство Penguin Books выпустило 42 книги, оно является лидером среди остальных издательств

## Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [9]:
query = """
SELECT 
    authors.author AS name_author,
    AVG(subq.rating_cnt) AS rating_cnt_avg,
    AVG(subq.rating_avg) AS rating_avg_avg
FROM
    books
INNER JOIN authors ON authors.author_id = books.author_id
INNER JOIN
    (
    SELECT 
        books.book_id AS book_id,
        books.title AS book_name,
        COUNT(ratings.rating) AS rating_cnt,
        AVG(ratings.rating) AS rating_avg
    FROM
        books
    INNER JOIN ratings ON ratings.book_id = books.book_id
    GROUP BY
        books.book_id
    ORDER BY
        rating_cnt DESC) AS subq
ON 
    books.title = subq.book_name
WHERE
    rating_cnt >= 50
GROUP BY
    name_author
ORDER BY
    rating_avg_avg DESC

    ;
"""

pd.io.sql.read_sql(query, con = engine)

name_author  rating_cnt_avg  \
0                          J.K. Rowling/Mary GrandPré            77.5   
1                   Markus Zusak/Cao Xuân Việt Khương            53.0   
2                                      J.R.R. Tolkien            81.0   
3                                   Louisa May Alcott            52.0   
4                                        Rick Riordan            62.0   
5                                     William Golding            71.0   
6                                       J.D. Salinger            86.0   
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce            57.0   
8   William Shakespeare/Paul Werstine/Barbara A. M...            66.0   
9                                           Dan Brown            71.5   
10                                         Lois Lowry            56.0   
11            George Orwell/Boris Grabnar/Peter Škerl            74.0   
12                                    Stephenie Meyer           160.0   
13                                     John Steinbeck            61.0   

    rating_avg_avg  
0         4.283844  
1         4.264151  
2         4.258446  
3         4.192308  
4         4.080645  
5         3.901408  
6         3.825581  
7         3.789474  
8         3.787879  
9         3.754540  
10        3.750000  
11        3.729730  
12        3.662500  
13        3.622951

Наибольшую среднюю оценку книг имеет J.K. Rowling/Mary GrandPré.

## Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [39]:
query = """
SELECT 
    AVG(subq.review_cnt) AS review_avg_from_user   
FROM
    (
    SELECT 
        reviews.username,
        COUNT(reviews.review_id) AS review_cnt
    FROM 
        reviews
    WHERE reviews.username in
        (
        SELECT 
            ratings.username AS username
        FROM 
            ratings
        GROUP BY
            ratings.username
        HAVING 
            COUNT(ratings.rating_id) >= 50) 
    GROUP BY
        reviews.username) AS subq
;
"""
pd.io.sql.read_sql(query, con = engine)

review_avg_from_user
0             24.222222

Среднее количество обзоров от пользователей с более чем 50 оценками составляет 24.22.